File: ruby/easy/convert_bst_to_greater_tree.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {TreeNode}
def convert_bst(root)
    return sum_up_bst(root, 0)[:root]
end

def sum_up_bst(root, higher_sum)
    return {sum: 0} unless root
    right = sum_up_bst(root.right, higher_sum)
    left = sum_up_bst(root.left, (higher_sum + root.val + right[:sum]))
    sum = root.val + left[:sum] + right[:sum]
    root.val += higher_sum + right[:sum]
    return { sum: sum, root: root }
end

File: javascript/easy/convertBSTToGreaterTree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
var convertBST = function(root) {
    convertToHigherTree(root, {sum: 0});
    return root;
};

var convertToHigherTree = function(root, obj) {
    if(root === null) return;
    convertToHigherTree(root.right, obj);
    obj.sum += root.val;
    root.val = obj.sum;
    convertToHigherTree(root.left, obj);   
};
